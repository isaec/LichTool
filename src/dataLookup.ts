// embedding raw is a performance optimization
// https://v8.dev/blog/cost-of-javascript-2019#json
import data from "@data/data.json?raw";
import filters from "@data/filters.json?raw";
import {
  Distances,
  Durations,
  Levels,
  Ranges,
  SchoolAbbreviations,
  Sources,
  DurationUnits,
  TimeObject,
  DurationObject,
  ComponentsObject,
  Abilities,
  ItemTypes,
  ItemRarity,
} from "@src/generalTypes";
import { DataGroup } from "./components/Renderer/types";
import { extractTypeFromUrl } from "./formatter";

export const isType =
  <T extends DataUnion>(type: string) =>
  (dataObj: DataUnion): dataObj is T =>
    extractTypeFromUrl(dataObj.id) === type;

export type IdData = {
  id: string;
  name: string;
  source: Sources;
};

export type RawData<T> = Omit<T, "id">;

export type RawDataSpell = RawData<DataSpell>;
export const isDataSpell = isType<DataSpell>("spell");
export type DataSpell = IdData & {
  level: Levels;
  school: SchoolAbbreviations;
  time: TimeObject[];
  range: { type: Ranges; distance?: { type: Distances; amount?: number } };
  components: ComponentsObject;
  duration: DurationObject[];
  entries: DataGroup;
  entriesHigherLevel?: DataGroup;
  page: number;
  savingThrow?: Abilities[];
  abilityCheck?: Abilities[];
  /**
   * M: Melee
   * R: Ranged
   * O: Other/Unknown
   */
  spellAttack?: Array<"M" | "R" | "O">;
  srd?: boolean;
  basicRules?: boolean;
  meta?: {
    ritual?: boolean;
  };
};

export type RawDataItem = RawData<DataItem>;
export const isDataItem = isType<DataItem>("item");
export type DataItem = IdData & {
  page: number;
  srd?: boolean;
  basicRules?: boolean;
  type?: ItemTypes;
  rarity?: ItemRarity;
  entries?: DataGroup;
  additionalEntries?: DataGroup;
};

export type RawConditionData = RawData<DataCondition>;
export const isDataCondition = isType<DataCondition>("condition");
export type DataCondition = IdData & {
  entries: DataGroup;
  page: number;
  srd?: boolean;
  basicRules?: boolean;
};

export type RawDiseaseData = RawData<DataDisease>;
export const isDataDisease = isType<DataDisease>("disease");
export type DataDisease = IdData & {
  entries: DataGroup;
  page: number;
  srd?: boolean;
};

export type RawStatusData = RawData<DataStatus>;
export const isDataStatus = isType<DataStatus>("status");
export type DataStatus = IdData & {
  name: string;
  entries: DataGroup;
  source: Sources;
  page: number;
  srd?: boolean;
  basicRules?: boolean;
};

export type DataUnion =
  | DataSpell
  | DataItem
  | DataCondition
  | DataDisease
  | DataStatus;

// data object collections

/** the array of all spells - id is always undefined */
export const dataArray = JSON.parse(data) as Array<DataUnion>;

/**
map from id to data.
*/
export const dataMap = new Map(
  dataArray.map((data) => [data.id, data])
) as ReadonlyMap<string, DataUnion>;

type ValueTypes = string | number | boolean;

/**
 * an array of all possible values for a filter
 */
export const filterMap = new Map<string, ValueTypes[]>(JSON.parse(filters));

/**
 * The keys of {@link filterMap} as an array
 */
export const filterKeys = [...filterMap.keys()];
