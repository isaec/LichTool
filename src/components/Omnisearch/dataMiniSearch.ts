import { spellArray } from "@src/dataLookup";
import MiniSearch from "minisearch";

// use the same minisearch for each search instance
const searchEngine = new MiniSearch({
  fields: ["name"],
  searchOptions: {
    fuzzy: 0.5,
    prefix: true,
    weights: {
      fuzzy: 0.65,
      prefix: 1,
    },
  },
});
searchEngine.addAll(spellArray);

export default searchEngine;
