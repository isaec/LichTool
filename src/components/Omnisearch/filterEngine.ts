import toast from "solid-toast";
import { dataMap, filterMap } from "@src/dataLookup";

export type FilterData = {
  key: string;
  value: string;
};
export type PopulatedFilter = FilterData & {
  use: true;
};
export type BlankFilter = Partial<FilterData> & {
  use: false;
};
export type Filter = BlankFilter | PopulatedFilter;

export const isBlank = (filter: Filter): filter is BlankFilter => !filter.use;
export const isPopulated = (filter: Filter): filter is PopulatedFilter =>
  filter.use;

/**
 * this is a cache of parsed filters, but only expensive filters will be cached
 */
const filterParseMap = new Map<string, ReturnType<typeof parseFilter>>();

const isOnlyDigits = /^\d+$/;
// flags m i and u aren't that useful, but are allowed
// i is the most useful of them
const isValidRegex = /^\/(.+)\/([miu]*)$/;
export const parseFilter = (
  filter: string
): boolean | number | RegExp | string | null => {
  if (filter === "true") return true;
  if (filter === "false") return false;

  if (isOnlyDigits.test(filter)) return parseInt(filter);
  if (isValidRegex.test(filter)) {
    // it would be faster for regex to read cache earlier and not need to test
    // but only checking cache for valid regex speeds up other cases
    if (filterParseMap.has(filter)) return filterParseMap.get(filter)!;
    const [, regex, flags] = isValidRegex.exec(filter)!;
    try {
      const val = new RegExp(regex, flags);
      filterParseMap.set(filter, val);
      return val;
    } catch (e) {
      // this prevents the immense cost of parsing invalid regex 1000x
      filterParseMap.set(filter, null);
      // this removes the invalid regex from the cache
      setTimeout(() => filterParseMap.delete(filter), 1000);
      // show that error once we finish rendering
      setTimeout(() => {
        toast.error((e as Error).message);
      }, 0);
      return null;
    }
  }
  return filter.toLowerCase();
};

// not in use right now...
const filterValueTransforms = new Map<string, Map<any, any>>([]);

/**
 * @param dataObj the data object to filter based on
 * @param filterObj the details of the filter to apply
 * @returns true if the data object matches the filter
 */
export const testFilter = (
  dataObj: Record<string, any>,
  filterObj: PopulatedFilter
) => {
  const filter = parseFilter(filterObj.value);
  if (filter === null) return false;
  const key = filterObj.key;
  const val = dataObj[key];
  if (filter instanceof RegExp) {
    return filter.test((val ?? "").toString());
  }
  if (typeof filter === "number" || typeof filter === "boolean")
    return val === filter;
  if (typeof filter === "string" && typeof val === "string") {
    const transformedValue = filterValueTransforms.get(key)?.get(val as any);
    if (transformedValue !== undefined) {
      return (
        transformedValue.toLowerCase().includes(filter) ||
        val.toLowerCase().includes(filter)
      );
    }
    return val.toLowerCase().includes(filter);
  }
  return false;
};

// this regex is used to match the filter keys in the url param
export const isParamFilter = /^f_/;

/**
 * Higher order function that returns a function to be used for filter execution
 *
 * If there are not filters, returns undefined
 */
export const executeFilters = (filters: PopulatedFilter[]) => {
  if (filters.length === 0) return undefined;
  return (result: { id: any }) => {
    const dataObj = dataMap.get(result.id)!;
    return !filters.some((filter) => !testFilter(dataObj, filter));
  };
};

export const filterIsValid = (filter: Filter) => {
  if (filter.key === undefined || filter.value === undefined) return false;
  if (filter.key === "") return false;
  if (filter.value === "") return false;
  // at this point the data is not literally missing
  if (!filterMap.has(filter.key)) return false;
  // we now know this could be a valid filter
  // regex on a valid key is valid
  if (isValidRegex.test(filter.value)) return true;
  // if it's a number, it's valid
  if (isOnlyDigits.test(filter.value)) return true;
  // if its a string which is a substring of possible values, it's valid
  const possibleValues = filterMap.get(filter.key)!;
  return possibleValues.some(
    (val) =>
      typeof val === "string" && val.toLowerCase().includes(filter.value!)
  );
};
