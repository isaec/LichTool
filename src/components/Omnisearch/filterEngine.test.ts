import { dataArray } from "@src/dataLookup";
import { describe, expect, it } from "vitest";
import dataMiniSearch from "./dataMiniSearch";
import { executeFilters, parseFilter, PopulatedFilter } from "./filterEngine";

describe("filterEngine", () => {
  const tr = (arr: {}[]) => arr.map((o) => ({ ...o, use: true }));
  const cases = [
    {
      query: "fireball",
      filters: tr([{ key: "name", value: "fireball" }]),
    },
    {
      filters: tr([
        {
          key: "name",
          value: `/mind\\b/i`,
        },
      ]),
    },
    {
      query: "m",
      filters: tr([
        {
          key: "level",
          value: "3",
        },
        {
          key: "name",
          value: "i",
        },
      ]),
    },
  ] as {
    query: string;
    filters: Array<PopulatedFilter>;
  }[];

  it.each(
    cases.map((c): [string, string, PopulatedFilter[]] => [
      c.query,
      JSON.stringify(c.filters),
      c.filters,
    ])
  )(
    "matches snapshot for search results given query %s, filter %s",
    (query, _filters, filters) => {
      if (query === undefined) {
        console.log(filters);
        expect(dataArray.filter(executeFilters(filters)!)).toMatchSnapshot();
      } else {
        expect(
          dataMiniSearch.search(query, {
            filter: executeFilters(filters),
          })
        ).toMatchSnapshot();
      }
    }
  );
});

describe("parseFilter", () => {
  describe.each([
    [
      "number",
      [
        ["1", 1],
        ["55", 55],
      ],
    ],
    [
      "boolean",
      [
        ["true", true],
        ["false", false],
      ],
    ],
    [
      "string",
      [
        ["fireball", "fireball"],
        ["fire blt", "fire blt"],
        ["123!", "123!"],
        ["5.5", "5.5"],
        ["/f", "/f"],
        [
          "/this looks like regex/ but its not!",
          "/this looks like regex/ but its not!",
        ],
      ],
    ],
    [
      "RegExp",
      [
        ["/f/", /f/],
        ["/^f/", /^f/],
        ["/something/i", /something/i],
      ],
    ],
  ] as Array<[string, Array<[string, ReturnType<typeof parseFilter>]>]>)(
    "parses %s into the correct literal",
    (type, cases) => {
      it.each(cases)(
        `parses %s into ${type} literal of "%s"`,
        (string, literal) => {
          expect(parseFilter(string)).toStrictEqual(literal);
        }
      );
    }
  );
});
