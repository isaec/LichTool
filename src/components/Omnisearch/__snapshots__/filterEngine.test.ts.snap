// Vitest Snapshot v1

exports[`filterEngine > matches snapshot for search results given query fireball, filter [{"key":"name","value":"fireball","use":true}] 1`] = `
[
  {
    "id": "spell_PHB-Fireball",
    "match": {
      "fireball": [
        "name",
      ],
    },
    "score": 8.225398227232713,
    "terms": [
      "fireball",
    ],
  },
  {
    "id": "spell_PHB-Delayed-Blast-Fireball",
    "match": {
      "fireball": [
        "name",
      ],
    },
    "score": 6.287348888500212,
    "terms": [
      "fireball",
    ],
  },
]
`;

exports[`filterEngine > matches snapshot for search results given query m, filter [{"key":"level","value":"3","use":true},{"key":"name","value":"i","use":true}] 1`] = `
[
  {
    "id": "spell_PHB-Major-Image",
    "match": {
      "major": [
        "name",
      ],
    },
    "score": 6.302528376091795,
    "terms": [
      "major",
    ],
  },
  {
    "id": "spell_PHB-Meld-into-Stone",
    "match": {
      "meld": [
        "name",
      ],
    },
    "score": 5.67903503998296,
    "terms": [
      "meld",
    ],
  },
  {
    "id": "spell_PHB-Mass-Healing-Word",
    "match": {
      "mass": [
        "name",
      ],
    },
    "score": 4.503687819691761,
    "terms": [
      "mass",
    ],
  },
  {
    "id": "spell_PHB-Dispel-Magic",
    "match": {
      "magic": [
        "name",
      ],
    },
    "score": 4.391636030244226,
    "terms": [
      "magic",
    ],
  },
  {
    "id": "spell_PHB-Magic-Circle",
    "match": {
      "magic": [
        "name",
      ],
    },
    "score": 4.391636030244226,
    "terms": [
      "magic",
    ],
  },
]
`;

exports[`filterEngine > matches snapshot for search results given query undefined, filter [{"key":"name","value":"/mind\\\\b/i","use":true}] 1`] = `
[
  {
    "areaTags": [
      "ST",
    ],
    "classes": {
      "fromClassList": [
        {
          "name": "Bard",
          "source": "PHB",
        },
        {
          "name": "Druid",
          "source": "PHB",
        },
        {
          "name": "Warlock",
          "source": "PHB",
        },
        {
          "name": "Wizard",
          "source": "PHB",
        },
      ],
    },
    "components": {
      "m": "a handful of clay, crystal, glass, or mineral spheres",
      "s": true,
      "v": true,
    },
    "damageInflict": [
      "psychic",
    ],
    "duration": [
      {
        "type": "instant",
      },
    ],
    "entries": [
      "You blast the mind of a creature that you can see within range, attempting to shatter its intellect and personality. The target takes {@damage 4d6} psychic damage and must make an Intelligence saving throw.",
      "On a failed save, the creature's Intelligence and Charisma scores become 1. The creature can't cast spells, activate magic items, understand language, or communicate in any intelligible way. The creature can, however, identify its friends, follow them, and even protect them.",
      "At the end of every 30 days, the creature can repeat its saving throw against this spell. If it succeeds on its saving throw, the spell ends.",
      "The spell can also be ended by {@spell greater restoration}, {@spell heal}, or {@spell wish}.",
    ],
    "id": "spell_PHB-Feeblemind",
    "level": 8,
    "miscTags": [
      "SGT",
    ],
    "name": "Feeblemind",
    "page": 239,
    "range": {
      "distance": {
        "amount": 150,
        "type": "feet",
      },
      "type": "point",
    },
    "savingThrow": [
      "intelligence",
    ],
    "school": "E",
    "source": "PHB",
    "srd": true,
    "time": [
      {
        "number": 1,
        "unit": "action",
      },
    ],
  },
  {
    "classes": {
      "fromClassList": [
        {
          "name": "Bard",
          "source": "PHB",
        },
        {
          "name": "Wizard",
          "source": "PHB",
        },
      ],
    },
    "components": {
      "s": true,
      "v": true,
    },
    "conditionImmune": [
      "charmed",
    ],
    "damageImmune": [
      "psychic",
    ],
    "duration": [
      {
        "duration": {
          "amount": 24,
          "type": "hour",
        },
        "type": "timed",
      },
    ],
    "entries": [
      "Until the spell ends, one willing creature you touch is immune to psychic damage, any effect that would sense its emotions or read its thoughts, divination spells, and the {@condition charmed} condition. The spell even foils {@spell wish} spells and spells or effects of similar power used to affect the target's mind or to gain information about the target.",
    ],
    "id": "spell_PHB-Mind-Blank",
    "level": 8,
    "name": "Mind Blank",
    "page": 259,
    "range": {
      "distance": {
        "type": "touch",
      },
      "type": "point",
    },
    "school": "A",
    "source": "PHB",
    "srd": true,
    "time": [
      {
        "number": 1,
        "unit": "action",
      },
    ],
  },
]
`;
